docs/design/sql/16-nested-fields-access.md

```diff
@@ -0,0 +1,359 @@
+Person>`, then the mapping will have these fields:
+
+- `__key BIGINT`
+- `this OBJECT` (hidden)
+- `name`, `birthDate`, ... (all fields of the `Person` class)
+
+This behavior remains unchanged, that is, even if there is a UDT for `Person`,
+`this` will have `OBJECT` type. 
+
+## ROW values
+
+The SQL standards specifies the ROW type. A ROW value contains an array of
+values, but doesn't contain field names. It is important to note that UDT values
+aren't ROW values. However, they can be converted both ways:
+
+- `TO_ROW` function converts an UDT value to `ROW` value
+
+- `CAST(rowExpression AS <UDT name>)`: converts a `ROW` value to a UDT value.
+  The ROW value must have the same number and types of fields as the UDT, in the
+  same order.
+
+When the ROW value is returned to the client, a
+`com.hazelcast.sql.impl.expression.RowValue` is returned to the java client.
+Note that it is a private API class - the reason is that STRUCT types aren't
+supported by JDBC. We could make this API public in our custom SQL API, but we
+consider JDBC to be the primary Java API. This is also what PG and Oracle do
+(TODO confirm Oracle). The user can reasonably only use `toString` on the value,
+which is what JDBC UI clients do, the `toString` will contain comma-separated
+list of all fields.
+
+The `ROW` type supports comparison operators that is different from how the
+comparison operator works with `OBJECT` type. It compares all field one by one,
+from left to right (this is important for operators other than `=`).
+
+The `ROW` type can also be assigned to another `ROW` type, provided they have
+the same number and types of fields. This is important with `ROW` literals, they
+can be used to create UDT values:
+
+```sql
+UPDATE m
+SET address=CAST(('street', 'city', 12345, 'Slovakia') AS AddressType
+WHERE ...;
+```
+
+### Semantics of comparison operators
+
+Comparison operators are: `=`, `!=`, `<`, `>`, `<=`, `>=`. `IS [NOT] DISTINCT
+FROM` isn't currently supported.
+
+When one operand of a comparison operator is ROW, the other operand must:
+- have a ROW type
+- have the same number of fields
+
+For individual fields of the ROW, we apply normal rules for comparison operator
+recursively. That is, `(value1, value2) = (value3, value4)` is equivalent to
+`value1 = value3 AND value2 = value4`, including the type conversion this
+involves.
+
+For example: 
+
+```
+(INT, INT) <cmp> (BIGINT, INT)   # ok, INT converted to BIGINT
+(INT, INT) <cmp> OBJECT  # reject ROW and OBJECT comparison
+(INT) <cmp> (VARCHAR) # ok, VARCHAR onverted to INT
+(INT) <cmp> (INT, INT)  # reject - different number of fields
+(INT, (INT, BIGINT)) <cmp> (TINYINT, (BIGINT, INT))  # ok, same rules for nested ROW
+```
+
+The comparison will be done from left to right. For evaluation, initial equal
+fields are ignored, and the first non-equal field determines the result. If all
+fields are equal, any field determines the result.
+
+### Issues with `CAST(field AS ROW)`
```

Comment:
This was a simple issue with the parser, default ROW TypeSpec declaration in the parser requires defining both field names and field types. This is massively impractical for end users since it leads to expressions like: 
`CAST (organization AS ROW(id BIGINT, name VARCHAR, office ROW(id BIGINT, name VARCHAR)))` which imo is completely impractical. We could redefine the `ROW` TypeSpec of course, but it would probably still not be very practical unless we allow field-less `ROW` TypeSpec, which itself might cause some problems with validation, hard to say what would be the side effects. 

https://github.com/apache/calcite/blob/main/core/src/main/codegen/templates/Parser.jj#L5710